pages/handbook/product-engineering/playbooks/code-structure.mdx

```diff
@@ -0,0 +1,178 @@
+---
+description: Learn how Langfuse's backend is structured and how to write code following our patterns and best practices.
+---
+
+# Backend Code Structure Guide
+
+This guide explains how Langfuse's backend is organized and how to write code that follows our established patterns. 
+
+## Architecture at a Glance
+
+Langfuse uses a **monorepo** structure with three main packages:
+
+- **web** - Next.js 14 application (UI + tRPC API + Public REST API)
+- **worker** - Express-based background job processor using BullMQ
+- **packages/shared** - Shared code, types, and utilities used by both web and worker
+
+### API Request Flow
+
+```
+┌─ Web (NextJs): tRPC API ────┐   ┌── Web (NextJs): Public API ─┐
+│                             │   │                             │
+│  HTTP Request               │   │  HTTP Request               │
+│      ↓                      │   │      ↓                      │
+│  tRPC Procedure             │   │  withMiddlewares +          │
+│  (protectedProjectProcedure)│   │  createAuthedProjectAPIRoute│
+│      ↓                      │   │      ↓                      │
+│  Service (business logic)   │   │  Service (business logic)   │
+│      ↓                      │   │      ↓                      │
+│  Prisma / ClickHouse        │   │  Prisma / ClickHouse        │
+│                             │   │                             │
+└─────────────────────────────┘   └─────────────────────────────┘
+                 ↓
+            [optional]: Publish to Redis BullMQ queue
+                 ↓
+┌─ Worker (Express): BullMQ Queue Job ────────────────────────┐
+│                                                             │
+│  BullMQ Queue Job                                           │
+│      ↓                                                      │
+│  Queue Processor (handles job)                              │
+│      ↓                                                      │
+│  Service (business logic)                                   │
+│      ↓                                                      │
+│  Prisma / ClickHouse                                        │
+│                                                             │
+└─────────────────────────────────────────────────────────────┘
+```
+We follow the layered architecture pattern:
+- Router Layer: HTTP Requests or BullMQ Job handlers
+- Service Layer: Contains all the business logic
+- Repository Layer: Prisma / ClickHouse
+
+## Directory Structure
+
+### Web Package (`/web/src/`)
```

Comment:
What I am missing: Where do I start developing a new feature? 

My assumption would be:

- start in web/src/features/new-feature-name
- all files go there
- ONLY add something to the shared package, if both worker and web need it